File: problems/0503_next-greater-element-ii/solution.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var nextGreaterElements = function (nums) {
    const n = nums.length;
    const result = new Array(n).fill(-1);
    const stack = [];
    for (let i = 0; i < 2 * n; i++) {
        const idx = i % n;
        while (stack.length > 0 && nums[stack[stack.length - 1]] < nums[idx]) {
            result[stack.pop()] = nums[idx];
        }
        if (i < n) {
            stack.push(idx);
        }
    }
    return result;
};

File: problems/0503_next-greater-element-ii/solution.go
func nextGreaterElements(nums []int) []int {
	n := len(nums)
	result := make([]int, n)
	for i := range result {
		result[i] = -1
	}
	stack := make([]int, 0, n)
	for i := 0; i < 2*n; i++ {
		idx := i % n
		for len(stack) > 0 && nums[stack[len(stack)-1]] < nums[idx] {
			result[stack[len(stack)-1]] = nums[idx]
			stack = stack[:len(stack)-1]
		}
		if i < n {
			stack = append(stack, idx)
		}
	}
	return result
}
